test-data/unit/check-python38.test

```diff
@@ -181,3 +181,122 @@ def f(p1: bytes, p2: float, /, p_or_kw: int) -> None:
+
+    y5 = (y6 := 3)
+    reveal_type(y5)  # E: Revealed type is 'builtins.int'
+    reveal_type(y6)  # E: Revealed type is 'builtins.int'
+
+    f(x=(y7 := 3))
+    reveal_type(y7)  # E: Revealed type is 'builtins.int'
+
+    reveal_type((lambda: (y8 := 3) and y8)())  # E: Revealed type is 'builtins.int'
+
+    y7 = 1.0  # E: Incompatible types in assignment (expression has type "float", variable has type "int")
+    if y7 := "x":  # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+        pass
+
+    # Just make sure we don't crash on this sort of thing.
+    if (NT := NamedTuple("NT", [("x", int)])):  # E: "int" not callable
+        z2: NT  # E: Invalid type "NT"
+
+    if (Alias := int):
+        z3: Alias  # E: Invalid type "Alias"
+
+    return (y8 := 3) + y8
+
+reveal_type(c)  # E: Revealed type is 'builtins.int'
+
+[builtins fixtures/f_string.pyi]
+
+[case testWalrusNewSemanal]
+# flags: --new-semantic-analyzer
+from typing import NamedTuple
+
+if (a := 2):
+    reveal_type(a)  # E: Revealed type is 'builtins.int'
+
+while (b := "x"):
+    reveal_type(b)  # E: Revealed type is 'builtins.str'
+
+def f(x: int = (c := 4)) -> int:
+    if (a := 2):
+        reveal_type(a)  # E: Revealed type is 'builtins.int'
+
+    while (b := "x"):
+        reveal_type(b)  # E: Revealed type is 'builtins.str'
+
+    x = (y := 1) + (z := 2)
+    reveal_type(x)  # E: Revealed type is 'builtins.int'
+    reveal_type(y)  # E: Revealed type is 'builtins.int'
+    reveal_type(z)  # E: Revealed type is 'builtins.int'
+
+    l = [y2 := 1, y2 + 2, y2 + 3]
+    reveal_type(y2)  # E: Revealed type is 'builtins.int'
+    reveal_type(l)  # E: Revealed type is 'builtins.list[builtins.int*]'
+
+    filtered_data = [y3 for x in l if (y3 := a) is not None]
+    reveal_type(filtered_data)  # E: Revealed type is 'builtins.list[builtins.int*]'
+    reveal_type(y3)  # E: Revealed type is 'builtins.int'
+
+    # https://www.python.org/dev/peps/pep-0572/#exceptional-cases
+    (y4 := 3)
+    reveal_type(y4)  # E: Revealed type is 'builtins.int'
+
+    y5 = (y6 := 3)
+    reveal_type(y5)  # E: Revealed type is 'builtins.int'
+    reveal_type(y6)  # E: Revealed type is 'builtins.int'
+
+    f(x=(y7 := 3))
+    reveal_type(y7)  # E: Revealed type is 'builtins.int'
+
+    reveal_type((lambda: (y8 := 3) and y8)())  # E: Revealed type is 'builtins.int'
+
+    y7 = 1.0  # E: Incompatible types in assignment (expression has type "float", variable has type "int")
+    if y7 := "x":  # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+        pass
+
+    # Just make sure we don't crash on this sort of thing.
+    if (NT := NamedTuple("NT", [("x", int)])):  # E: "int" not callable
+        z2: NT  # E: Invalid type "NT"
+
+    if (Alias := int):
+        z3: Alias  # E: Invalid type "Alias"
+
+    return (y8 := 3) + y8
+
+reveal_type(c)  # E: Revealed type is 'builtins.int'
+
+[builtins fixtures/f_string.pyi]
```

Comment:
I would like to see more checker related tests here, namely:
* Interactions with final names/attributes (see the comment above)
* Interactions with binder (this will probably work as is, but without a test factoring out binder code from `check_assignment()` to containing method will cause the same problem as with `Final` now)
* Interactions with partial types and deferred nodes (mostly for the same reason as above)

Reply:
Adding this now. I'm omitting final attributes because only names are allowed in the LHS of :=.

```
$ python -c 'if self.x := 2: pass'
  File "<string>", line 1
SyntaxError: cannot use named assignment with attribute
```
Also, what would be a good test for a deferred node? Perhaps a variable assigned in an assignment expression with a type that is only defined later in the file?

Reply:
> Perhaps a variable assigned in an assignment expression with a type that is only defined later in the file?

Yes, r.h.s. can be an inferred attribute of some class that is initialized with a function call.